src/CoreWCF.Primitives/src/CoreWCF/Channels/ByteStreamMessageEncoder.cs

```diff
@@ -79,7 +75,11 @@ public override Message ReadMessage(ArraySegment<byte> buffer, BufferManager buf
             //    TD.ByteStreamMessageDecodingStart();
             //}
 
-            var messageData = new ByteStreamBufferedMessageData(buffer, bufferManager);
+            int bufferLength = (int)buffer.Length;
+            byte[] bytes = bufferManager.TakeBuffer(bufferLength);
+            buffer.CopyTo(bytes.AsSpan(0, bufferLength));
+
```

Comment:
This is going to reduce performance as we're copying the ReadOnlySequence into an allocated buffer. Can we do anything in ByteStreamBufferedMessageData so use the ReadOnlySequence instead?

Reply:
I looked at the code and I believe we can just keep the ReadOnlySequence. There's some changes needed in ByteStreamMessage. Where it calls writer.WriteBase64, we should check if writer is an XmlByteStreamWriter (which we own) and case to that. Have a new overload on XmlByteStreamWriter which takes a ReadOnlySequence and have a ReadOnlySequence specific implementation. All it does is call Stream.Write. ReadOnlySequence has an enumerator which emits ReadOnlyMemory. The problem here is netstandard2.0 doesn't have a Stream.Write overload which can rake a ReadOnlyMemory. There is a MemoryMarshal helper method which can try to get you the underlying buffer which you can pass to Stream.Write, but if it can't (eg it's backed by native memory), then you need to allocate a buffer and copy it. I have code to do this in the NetNamedPipe implementation so if you are interested in getting all this working, I can point you at the code.

Reply:
@mconnew this is now done